### PR TITLE
Self-dependent structs.

### PR DESCRIPTION
**Pull request by mzhurovich**

### CODE REVIEW

File: TypeSystem/Reflection/schema.h [hunk 1]
```diff
@@ -227,18 +228,21 @@ struct StructSchema {
...
     const uint64_t struct_type_prefix = TypePrefix(struct_type_id);
     assert(struct_type_prefix == TYPEID_STRUCT_PREFIX);
-    RecursiveListStructDependencies(struct_type_id, result, false);
+    RecursiveListStructDependencies(struct_type_id, result, TypeID::INVALID_TYPE, false);
     return result;
   }
 
   void RecursiveListStructDependencies(const TypeID type_id,
                                        std::vector<TypeID>& dependency_list,
+                                       const TypeID containig_struct_type_id,
```
Comment (dkorolev) [inline TypeSystem/Reflection/schema.h]: Typo, `containing_*`?

File: TypeSystem/Reflection/types.h [hunk 2]
```diff
@@ -127,8 +129,11 @@ inline uint64_t ROL64(const TypeID type_id, size_t nbits) {
   return current::ROL64(static_cast<uint64_t>(type_id), nbits);
 }
 
-inline TypeID CalculateTypeID(const ReflectedType_Struct& s) {
+inline TypeID CalculateTypeID(const ReflectedType_Struct& s, bool is_incomplete = false) {
   uint64_t hash = bricks::CRC32(s.name);
+  if (is_incomplete) {
+    return static_cast<TypeID>(TYPEID_INCOMPLETE_TYPE + hash);
+  }
```
Comment (dkorolev) [inline TypeSystem/Reflection/types.h]: Still suspicious about `i + 8` and `i + 16` below. Can they be some `i + 19` and `i + 29` instead? :)

File: TypeSystem/Reflection/types.h [hunk 3]
```diff
@@ -127,8 +129,11 @@ inline uint64_t ROL64(const TypeID type_id, size_t nbits) {
   return current::ROL64(static_cast<uint64_t>(type_id), nbits);
 }
 
-inline TypeID CalculateTypeID(const ReflectedType_Struct& s) {
+inline TypeID CalculateTypeID(const ReflectedType_Struct& s, bool is_incomplete = false) {
   uint64_t hash = bricks::CRC32(s.name);
+  if (is_incomplete) {
+    return static_cast<TypeID>(TYPEID_INCOMPLETE_TYPE + hash);
```
Comment (dkorolev) [inline TypeSystem/Reflection/types.h]: Perhaps still hash+rol+xor the names of fields too, not just the name of the structure?

File: TypeSystem/Reflection/types.h [hunk 4]
```diff
@@ -45,22 +45,24 @@ namespace current {
...
 
 constexpr uint64_t TYPEID_TYPE_RANGE  = static_cast<uint64_t>(1e16);
-constexpr uint64_t TYPEID_BASIC_TYPE  = TYPEID_TYPE_RANGE * TYPEID_BASIC_PREFIX;
-constexpr uint64_t TYPEID_ENUM_TYPE   = TYPEID_TYPE_RANGE * TYPEID_ENUM_PREFIX;
-constexpr uint64_t TYPEID_STRUCT_TYPE = TYPEID_TYPE_RANGE * TYPEID_STRUCT_PREFIX;
-constexpr uint64_t TYPEID_VECTOR_TYPE = TYPEID_TYPE_RANGE * TYPEID_VECTOR_PREFIX;
-constexpr uint64_t TYPEID_SET_TYPE    = TYPEID_TYPE_RANGE * TYPEID_SET_PREFIX;
-constexpr uint64_t TYPEID_PAIR_TYPE   = TYPEID_TYPE_RANGE * TYPEID_PAIR_PREFIX;
-constexpr uint64_t TYPEID_MAP_TYPE    = TYPEID_TYPE_RANGE * TYPEID_MAP_PREFIX;
+constexpr uint64_t TYPEID_INCOMPLETE_TYPE  = TYPEID_TYPE_RANGE * TYPEID_INCOMPLETE_PREFIX;
```
Comment (dkorolev) [inline TypeSystem/Reflection/types.h]: Perhaps `INCOMPLETE_STRUCT` would be a better name?

File: TypeSystem/Reflection/schema.h (hunk 1 above)
Comment (mzhurovich) [inline TypeSystem/Reflection/schema.h]: :+1:

File: TypeSystem/Reflection/types.h (hunk 4 above)
Comment (mzhurovich) [inline TypeSystem/Reflection/types.h]: :+1:

File: TypeSystem/Reflection/types.h (hunk 2 above)
Comment (mzhurovich) [inline TypeSystem/Reflection/types.h]: Ok! Decided to change other constants too :)

File: TypeSystem/Reflection/types.h (hunk 3 above)
Comment (mzhurovich) [inline TypeSystem/Reflection/types.h]: Good idea!

**Comment by dkorolev** [comment]
Superseded by https://github.com/c5t/Current/pull/269.

**Pull request closed without merging**